File: MyLeetCode/Leetcode/Closest Binary Search Tree Value II.cpp
/**
 * The idea is to compare the predecessors and successors of the closest node to the target, 
 * we can use two stacks to track the predecessors and successors, 
 * then like what we do in merge sort, we compare and pick the closest one to the target and put it to the result list.
 * As we know, inorder traversal gives us sorted predecessors, whereas reverse-inorder traversal gives us sorted successors.
 * We can use iterative inorder traversal rather than recursion, but to keep the code clean, here is the recursion version.
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void inorder(TreeNode *root, stack<int> &stk, double target, int reverse)
    {
        if (!root) return;

        inorder(reverse ? root->right : root->left, stk, target, reverse);

        if ((!reverse && root->val > target)
            || (reverse && root->val <= target))
            return;

        stk.push(root->val);

        inorder(reverse ? root->left : root->right, stk, target, reverse);
    }

    vector<int> closestKValues(TreeNode* root, double target, int k) {
        
        vector<int> res;
        if (!root || !k) return res;
        stack<int> min_stk, max_stk;
        inorder(root, min_stk, target, 0);
        inorder(root, max_stk, target, 1);
        while (k--)
        {
            if (min_stk.empty())
            {
                res.push_back(max_stk.top());
                max_stk.pop();
            }
            else if (max_stk.empty())
            {
                res.push_back(min_stk.top());
                min_stk.pop();
            }
            else
            {
                if (abs(min_stk.top()-target) < abs(max_stk.top()-target))
                {
                    res.push_back(min_stk.top());
                    min_stk.pop();
                }
                else
                {
                    res.push_back(max_stk.top());
                    max_stk.pop();
                }
            }
        }
        return res;
    }
};

File: MyLeetCode/python/Closest Binary Search Tree Value II.py

###
# Tree Inorder
# Time Complexit: O(h+k)
# Space Complexity: O(h)
###
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def closestKValues(self, root, target, k):
        """
        :type root: TreeNode
        :type target: float
        :type k: int
        :rtype: List[int]
        """
        if not root:
            return []

        pred = self.init_pred(root, target)
        succ = self.init_succ(root, target)
        result = []
        print(pred)
        print(succ)
        while k:
            if not pred:
                result.append(self.get_next_succ(succ))
            elif not succ:
                result.append(self.get_next_pred(pred))
            else:
                if abs(pred[-1].val - target) < abs(succ[-1].val - target):
                    result.append(self.get_next_pred(pred))
                else:
                    result.append(self.get_next_succ(succ))
            k -= 1
        return result

    def init_pred(self, root, target):
        pred = []
        cur = root
        while cur:
            if cur.val <= target:
                pred.append(cur)
                cur = cur.right
            else:
                cur = cur.left
        return pred

    def init_succ(self, root, target):
        succ = []
        cur = root
        while cur:
            if cur.val > target:
                succ.append(cur)
                cur = cur.left
            else:
                cur = cur.right
        return succ

    def get_next_pred(self, pred):
        top = pred.pop()
        cur = top.left
        while cur:
            pred.append(cur)
            cur = cur.right
        return top.val

    def get_next_succ(self, succ):
        top = succ.pop()
        cur = top.right
        while cur:
            succ.append(cur)
            cur = cur.left
        return top.val


